### PR TITLE
Add tests for `verifyTx` and `verifyHeader`

### LINKED ISSUE #1305: Add more tests for header and transaction verification
There are a lot of subtle error cases there (`verifyHeader`, `verifyTx` and their inner functions) that are not tested at all at the moment (we're only going the happy path mostly), but they're really important at the same time, each of these checks has a purpose and we need to test them.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1305 .
Related #1292 .

### CODE REVIEW

File: pkg/core/blockchain_test.go [hunk 1]
```diff
@@ -144,6 +150,114 @@ func TestGetBlock(t *testing.T) {
...
+		Scopes:  transaction.CalledByEntry,
+	}}
+	tx.NetworkFee = int64(io.GetVarSize(tx)+200 /* witness */) * bc.FeePerByte()
+	tx.NetworkFee += 1_000_000 // verification cost
+	return tx
+}
+
+func TestVerifyTx(t *testing.T) {
+	bc := newTestChain(t)
+	privs := make([]*keys.PrivateKey, 2)
```
Comment (roman-khimov) [inline pkg/core/blockchain_test.go:195]: Do we really need them as a separate thing?

File: pkg/core/blockchain_test.go [hunk 2]
```diff
@@ -144,6 +150,114 @@ func TestGetBlock(t *testing.T) {
...
+	accs := make([]*wallet.Account, 2)
+	for i := range privs {
+		var err error
+		privs[i], err = keys.NewPrivateKey()
+		require.NoError(t, err)
+		accs[i], err = wallet.NewAccountFromWIF(privs[i].WIF())
```
Comment (roman-khimov) [inline pkg/core/blockchain_test.go:201]: `wallet.NewAccount()` and then get everything from this account?

File: pkg/core/blockchain.go [hunk 3]
```diff
@@ -1179,15 +1179,21 @@ func (bc *Blockchain) ApplyPolicyToTxSet(txes []*transaction.Transaction) []*tra
 	return txes
 }
 
+var (
+	errHdrHashMismatch     = errors.New("previous header hash doesn't match")
+	errHdrIndexMismatch    = errors.New("previous header index doesn't match")
+	errHdrInvalidTimestamp = errors.New("block is not newer than the previous one")
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1185]: export?

File: pkg/core/blockchain.go [hunk 4]
```diff
@@ -1192,11 +1192,19 @@ func (bc *Blockchain) verifyHeader(currHeader, prevHeader *block.Header) error {
 	return bc.verifyHeaderWitnesses(currHeader, prevHeader)
 }
 
+var (
+	errTxExpired         = errors.New("transaction has expired")
+	errInsufficientFunds = errors.New("insufficient funds")
+	errSmallNetworkFee   = errors.New("too small network fee")
+	errTxTooBig          = errors.New("too big transaction")
+	errMemPoolConflict   = errors.New("invalid transaction due to conflicts with the memory pool")
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/blockchain.go:1206]: They are already useful in tests.

File: pkg/core/blockchain.go (hunk 3 above)
Comment (fyrchik) [reply pkg/core/blockchain.go:1185]: Fixed.

File: pkg/core/blockchain.go [hunk 5]
```diff
@@ -1430,15 +1430,33 @@ func ScriptFromWitness(hash util.Uint160, witness *transaction.Witness) ([]byte,
...
-		return err
+	var offset int
+	var initMD *manifest.Method
+	verification := witness.VerificationScript
+	if len(verification) != 0 {
+		if witness.ScriptHash() != hash {
+			return ErrWitnessHashMismatch
+		}
+	} else {
+		cs, err := interopCtx.DAO.GetContractState(hash)
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1451]: It's a separate non-preview3-compatible feature, it shouldn't be a part of this PR.

Comment (fyrchik) [reply pkg/core/blockchain.go:1451]: Are you sure? It was 4 months ago neo-project/neo#1481

Comment (roman-khimov) [reply pkg/core/blockchain.go:1451]: Let me check, I was thinking more of neo-project/neo#1800.

Comment (roman-khimov) [reply pkg/core/blockchain.go:1450]: Yeah, this part is really preview3, but it should substitute `ScriptFromWitness()` then.

Comment (fyrchik) [reply pkg/core/blockchain.go:1450]: `ScriptFromWitness` has not necessary context, and its signature will become no so understandable.
I am thinking about making `verifyHashAgainstScript` public.

Comment (roman-khimov) [reply pkg/core/blockchain.go:1450]: That's a good option too, simplifying consensus a bit along the way.

Comment (fyrchik) [reply pkg/core/blockchain.go:1450]: I'd postpone it until after release, as it requires some refactoring of both `core` and `consensus` to do right.

**Pull request merged**